fili-core/src/main/java/com/yahoo/bard/webservice/web/handlers/EtagCacheRequestHandler.java

```diff
@@ -0,0 +1,174 @@
+    public static final Meter CACHE_HITS = REGISTRY.meter("queries.meter.cache.hits");
+    public static final Meter CACHE_POTENTIAL_HITS = REGISTRY.meter("queries.meter.cache.potential_hits");
+    public static final Meter CACHE_MISSES = REGISTRY.meter("queries.meter.cache.misses");
+    public static final Meter CACHE_REQUESTS = REGISTRY.meter("queries.meter.cache.total");
+
+    protected final @NotNull DataRequestHandler next;
+    protected final @NotNull TupleDataCache<String, Long, String> dataCache;
+    protected final @NotNull QuerySigningService<Long> querySigningService;
+
+    /**
+     * Build a Cache request handler.
+     *
+     * @param next  The next handler in the chain
+     * @param dataCache  The cache instance
+     * @param querySigningService  The service to generate query signatures
+     * @param mapper  The mapper for all JSON processing
+     */
+    @SuppressWarnings("unchecked")
+    public EtagCacheRequestHandler(
+            DataRequestHandler next,
+            @NotNull DataCache<?> dataCache,
+            QuerySigningService<?> querySigningService,
+            ObjectMapper mapper
+    ) {
+        super(mapper);
+        this.next = next;
+        this.dataCache = (TupleDataCache<String, Long, String>) dataCache;
+        this.querySigningService = (QuerySigningService<Long>) querySigningService;
+    }
+
+    @Override
+    public boolean handleRequest(
+            final RequestContext context,
+            final DataApiRequest request,
+            final DruidAggregationQuery<?> druidQuery,
+            final ResponseProcessor response
+    ) {
+        ResponseProcessor nextResponse = response;
+
+        String cacheKey = null;
+        try {
+            cacheKey = getKey(druidQuery);
+
+            if (context.isReadCache()) {
+                final TupleDataCache.DataEntry<String, Long, String> cacheEntry = dataCache.get(cacheKey);
+                CACHE_REQUESTS.mark(1);
+
+                if (cacheEntry != null) {
+                    context
+                            .getHeaders()
+                            .putSingle(
+                                    DruidJsonResponseContentKeys.ETAG.getName(),
+                                    mapper
+                                            .readTree(cacheEntry.getValue())
+                                            .get(DruidJsonResponseContentKeys.ETAG.getName())
+                                            .asText()
+                            );
+                    // Make sure that if the optional return value is empty, the statement always evaluates to false
+                    // Metadata type needs to be int.
+                    if (
+                            querySigningService.getSegmentSetId(druidQuery)
+                                    .map(id -> Objects.equals(cacheEntry.getMeta(), id))
+                                    .orElse(false)
+                            ) {
```

Comment:
Why are we looking at the locally-computed signature for this? Why not use _only_ the etags?